# The basics of linear regression

### Through a worked example

In [4]:
import numpy as np
import math

Here we have 11 data points in two dimensions $X$ and $Y$.
Our goal is to illustrate the construction of simple linear regression model by going through every step of the process covered until Week 4.

Familiarise yourself with the plot below.

![Alt text](./regression.png)


To get started, let's instantiate the eleven data points above as numpy arrays. 

Numpy arrays are better than native lists because numpy objects have mathematical methods we can use easily, like the mean we will use below.

In [5]:
x = np.array([0,1,1,2,3,3,4,5,6,6,7])
y = np.array([2,3,4,4,4,5,5,5,5,6,6])

Notice that the $(x,y)$ points correspond to the ones in the figure: (0,2), (1,3), ..., (7,6). 

Now lets instantiate a variable `n` with the number of data points.
We will also make sure our two numpy arrays are of the same size.

In [6]:
n = len(x)
if len(x) != len(y):
    print("wrong data, check.")

Our goal is to get a linear regression model, of the form:

$\hat{y} = \hat{\beta}_0 + \hat{\beta}_1x$

In other words, we want to compute $\beta_0$ and $\beta_1$
As discussed in the lectures, there are formulas to get these coefficients directly:

$$
\hat{\beta}_1 = \frac{\sum_{i=1}^n (x_i - \bar{x})(y_i - \bar{y})}{\sum_{i=1}^n (x_i - \bar{x})^2 }
$$

and

$$
\hat{\beta}_0 = \bar{y} - \hat{\beta}_1 \bar{x}
$$

Let's start by computing the mean of $x$ and $y$, which we will denote  by $\bar{x}$ and $\bar{y}$

In [7]:
mean_x = np.mean(x)
mean_y = np.mean(y)

Now lets compute the numerator and denominator in the formula for $\beta_1$, we will call them `v1` and `v2` respectively.

In [8]:
v1 = 0 # numerator for beta_1
v2 = 0 # denominator for beta_1

for i in range(len(x)):
    v1 += (x[i] - mean_x) * (y[i] - mean_y) 
    v2 += (x[i] - mean_x)**2


Now we are ready to compute $\beta_0$ and $\beta_1$

In [9]:
beta_1 = round(v1/v2, 2)
beta_0 = round(mean_y - beta_1 * mean_x, 2)

In [10]:
beta_0

2.83

In [11]:
beta_1

0.47

So our linear regression model is:

$ \hat{y} = 2.83 + 0.47*x $

Let's define a function that implements our model

In [13]:

def my_model_y(x):
    return beta_0 + beta_1*x

Now we can obtain the estimated values of $y$, i.e. $\hat{y}$ using our new model.

In [14]:
hat_y = np.array([my_model_y(el) for el in x])
hat_y

array([2.83, 3.3 , 3.3 , 3.77, 4.24, 4.24, 4.71, 5.18, 5.65, 5.65, 6.12])

In [15]:
# let's compute residuals and
# see correction on the slide

residuals = y - hat_y # the "palitos"
rss = round(sum([el**2 for el in residuals]), 2)

In [16]:
rss

2.63

Now, to compute the standard errors for the coefficients $\beta_0$ and $\beta_1$ we need to estimate $\sigma$, the true variance in the real system $Y$

In [17]:
sigma_estimate = round(math.sqrt(rss/(n-2)),2)
sigma_estimate

0.54

In [20]:
error_beta_1 = round(math.sqrt(sigma_estimate**2/v2),2)
# note that what we computed in v2, corresponds to what
# is in the denominator for the formula of standard error for beta_1
error_beta_1

0.07

In [21]:
error_beta_0 = round(math.sqrt(sigma_estimate**2 * (1/n + (mean_x**2/v2))),2)
error_beta_0

0.3

In [24]:

int_conf_beta_1 = [round(beta_1 - error_beta_1*2,2),  round(beta_1 + error_beta_1*2,2)]
int_conf_beta_1

[0.33, 0.61]

In [50]:
beta_1

0.47